### PR TITLE
Adds features to process constraint 'valid_values' and generate float and symbol from ion schema file.

### PR DESCRIPTION
**Pull request by linlin-s**
**Description of changes:**

_IonSchemaUtilities.java_
- parseConstraints:
   Changed the method of getting data from constraints when `min` or `max` provided.
- checkValidRange: 
   This method is used for checking whether the constraint `valid_value` has annotation `range`. If there is annotation  `range`, this method will return an IonList which contains the range of `valid_value`. This method will be used as an additional specification in the process of constructing data.
- parseValidValue:
  This method is used for parsing the constraint `valid_value` which only provided a list of IonValue without the annotation range. This method will be used in the first step of writing data. If provided ISL struct contains this constraint, the write-data-to-file process will not call other constructing data methods.

_WriteRandomIonValues.java_
- writeDataToFile:
  Added a step to check if a list of valid value has been provided in the constraint struct. If there is provided valid value, the writer will write the randomly selected data from the `valid_value` list to the file.
- constrcutFloat:
  This method is used for constructing double value. If the constraint `valid_value: range::[<number>]` provided, this method will construct value based on the specified valid range, else this method will construct double value randomly.
- constrcutInt:
  Added steps to process the constraint `valid_value: range::[<number>]`.

**Test:**

_DataGeneratorTest.java_
- Added test for generating float.
- Added test for generating float with specified `valid_value` range.
- Added test for generating symbol.
- Added test for generating string.
- Added test for generating int with specified `valid_value` range.

 
By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: src/com/amazon/ion/benchmark/IonSchemaUtilities.java
Comment (desaikd) [inline src/com/amazon/ion/benchmark/IonSchemaUtilities.java:178]: The method name here doesn't seem to identify with what it is doing. Maybe a name like `getValidValuesAsRange()` would work.

File: src/com/amazon/ion/benchmark/IonSchemaUtilities.java
Comment (desaikd) [inline src/com/amazon/ion/benchmark/IonSchemaUtilities.java:49]: As per Ion schema specification `valid_values` is correct the constraint name. 
Reference : https://amzn.github.io/ion-schema/docs/spec.html#valid_values

File: src/com/amazon/ion/benchmark/IonSchemaUtilities.java
Comment (desaikd) [inline src/com/amazon/ion/benchmark/IonSchemaUtilities.java:189]: Can we add a check for `range` here? This doesn't seem to check the annotation `range::`. There could be invalid annotations `range::`, `abc::` etc as Ion schema allows for open content.

File: src/com/amazon/ion/benchmark/IonSchemaUtilities.java
Comment (desaikd) [inline src/com/amazon/ion/benchmark/IonSchemaUtilities.java:200]: Does this mean it only parses for non range `valid_values`?  `without annotation range` makes it difficult to understand teh statement. Maybe add a comment as below would help:
```suggestion
     * This method only parses for `valid_values` that doesn't contain ranges.
```

File: tst/com/amazon/ion/benchmark/testFloatValidValue.isl
Comment (desaikd) [inline tst/com/amazon/ion/benchmark/testFloatValidValue.isl:6]: I am unsure how this test works. But Are we testing the generated data by validating it against the schema?
If yes, how does this test pass? Ion schema would not be able to understand `valid_value` constraint as it names as `valid_values`.

**Review by desaikd — COMMENTED** [review]
This is coming along well! There are some concerning changes I have addressed below. 
Mainly, are we testing the the generated data against the schema? Because `valid_value` is not there in the Ion schema grammar but `valid_values` is. Hence, seems like these tests would pass by ignoring the `valid_value` constraint as open content.

File: src/com/amazon/ion/benchmark/IonSchemaUtilities.java
Comment (linlin-s) [reply src/com/amazon/ion/benchmark/IonSchemaUtilities.java:178]: > The method name here doesn't seem to identify with what it is doing. Maybe a name like `getValidValuesAsRange()` would work.

Sounds good.

File: src/com/amazon/ion/benchmark/IonSchemaUtilities.java
Comment (linlin-s) [reply src/com/amazon/ion/benchmark/IonSchemaUtilities.java:189]: > Can we add a check for `range` here? This doesn't seem to check the annotation `range::`. There could be invalid annotations `range::`, `abc::` etc as Ion schema allows for open content.

Yes, I assumed that the `range::` is the only possibility. I'll add it in the following commit.

File: tst/com/amazon/ion/benchmark/testFloatValidValue.isl
Comment (linlin-s) [reply tst/com/amazon/ion/benchmark/testFloatValidValue.isl:6]: > I am unsure how this test works. But Are we testing the generated data by validating it against the schema? Ion schema would not be able to understand `valid_value` constraint as it names as `valid_values` and would consider it open content.

Yes, the first step of testing is to load the ion schema file and then test the generated data by validating it against the schema. If the grammar isn't right it suppose to throw errors. I'll double check the test method and make edition in the follow commit.

File: src/com/amazon/ion/benchmark/IonSchemaUtilities.java
Comment (tgregg) [inline src/com/amazon/ion/benchmark/IonSchemaUtilities.java:200]: paring -> parsing

File: src/com/amazon/ion/benchmark/IonSchemaUtilities.java
Comment (tgregg) [inline src/com/amazon/ion/benchmark/IonSchemaUtilities.java:207]: We can reuse one IonSystem and one IonLoader across all methods in the library.

File: src/com/amazon/ion/benchmark/WriteRandomIonValues.java
Comment (tgregg) [inline src/com/amazon/ion/benchmark/WriteRandomIonValues.java]: Consider restructuring this to avoid having to call `parseValidValue` twice.

File: src/com/amazon/ion/benchmark/WriteRandomIonValues.java
Comment (tgregg) [inline src/com/amazon/ion/benchmark/WriteRandomIonValues.java:364]: What if there's both a regex and a code point length constraint?

File: src/com/amazon/ion/benchmark/WriteRandomIonValues.java
Comment (tgregg) [inline src/com/amazon/ion/benchmark/WriteRandomIonValues.java]: Consider restructuring this to avoid having to call `checkValidRange` twice.

File: src/com/amazon/ion/benchmark/WriteRandomIonValues.java
Comment (tgregg) [inline src/com/amazon/ion/benchmark/WriteRandomIonValues.java:449]: We can achieve this behavior fairly easily without taking a dependency on `org.apache.commons.math3.random.RandomDataGenerator`. I recommend doing that unless there's another good reason for this dependency.

Comment (linlin-s) [reply src/com/amazon/ion/benchmark/WriteRandomIonValues.java]: > We can achieve this behavior fairly easily without taking a dependency on `org.apache.commons.math3.random.RandomDataGenerator`. I recommend doing that unless there's another good reason for this dependency.

Applied this suggestion in the following commit.

File: src/com/amazon/ion/benchmark/WriteRandomIonValues.java
Comment (linlin-s) [reply src/com/amazon/ion/benchmark/WriteRandomIonValues.java]: > What if there's both a regex and a code point length constraint?

In the following commit, the logic of constructing string will be categorized into three conditions:
1. ISL contains both `regex` and `codepoint_length`:
   First, generating string based on the provided `regex` then check whether the generated string aligned with the constraint `codepoint_length`. If not, the method will keep generating until the generated string matches both constraints. It might take a while to generate the string which matches two constraints at the same time or the provided constraints are conflict with each other (e.g. regex: "^B[0-9]{9}$", codepoint_length: range:: [1, 3]). Under this consideration, preset the times of implementing generating string step at most 10. When the iterations hit 10 times but not find any matches string, it will write null value.
2. Only contain one of the constraints:
   Construct the string conformed with the provided constraint.
3. No constraint:
   Under this condition, the method will construct string randomly with random Unicode codepoint and `codepoint_length`.

Note:
1. The upper bound of iteration time requires further discussion.
2. The unit test cannot be implemented until this issue be closed

File: src/com/amazon/ion/benchmark/WriteRandomIonValues.java
Comment (linlin-s) [reply src/com/amazon/ion/benchmark/WriteRandomIonValues.java]: > Consider restructuring this to avoid having to call `parseValidValue` twice.

Updated in the following commit.

File: src/com/amazon/ion/benchmark/IonSchemaUtilities.java
Comment (linlin-s) [reply src/com/amazon/ion/benchmark/IonSchemaUtilities.java]: > We can reuse one IonSystem and one IonLoader across all methods in the library.

Updated in the following commit.

File: src/com/amazon/ion/benchmark/IonSchemaUtilities.java
Comment (linlin-s) [reply src/com/amazon/ion/benchmark/IonSchemaUtilities.java]: Yes, this method could be used by other constraints which contains range too. All methods with the functionalities of understanding ion schema will be refactored in the next PR and refactoring this method is part of the plan. There will be a class QuantifiableConstraint to handle all constraints which contain range.

File: src/com/amazon/ion/benchmark/IonSchemaUtilities.java
Comment (linlin-s) [reply src/com/amazon/ion/benchmark/IonSchemaUtilities.java]: Yes this method only parse `valid_values: [<VALUE>... ] `. I'll update all typo of `valid_values` in the following commit.

File: src/com/amazon/ion/benchmark/WriteRandomIonValues.java
Comment (linlin-s) [reply src/com/amazon/ion/benchmark/WriteRandomIonValues.java]: The statement will be updated in the next commit.

File: src/com/amazon/ion/benchmark/WriteRandomIonValues.java
Comment (linlin-s) [reply src/com/amazon/ion/benchmark/WriteRandomIonValues.java]: > 

Thanks for suggestions. This is also part of the consideration of refactoring. The refactoring will be contained in the next PR.

File: src/com/amazon/ion/benchmark/WriteRandomIonValues.java
Comment (tgregg) [inline src/com/amazon/ion/benchmark/WriteRandomIonValues.java:473]: You already have a `Random` instance, and should probably use that here instead of `Math.random()` (which creates its own `Random` instance). That should allow you to use `nextLong()` to avoid the cast.

File: tst/com/amazon/ion/benchmark/DataGeneratorTest.java
Comment (tgregg) [inline tst/com/amazon/ion/benchmark/DataGeneratorTest.java:229]: floats -> symbols

This same copy/paste error occurs on other JavaDoc comments on the test methods in this file.

**Review by tgregg — APPROVED** [review]
Two minor comments. Approved pending resolution of those and approval by @desaikd.

File: tst/com/amazon/ion/benchmark/DataGeneratorTest.java
Comment (linlin-s) [reply tst/com/amazon/ion/benchmark/DataGeneratorTest.java]: > floats -> symbols
> 
> This same copy/paste error occurs on other JavaDoc comments on the test methods in this file.

Thanks, the updates will be included in the next commit.

File: src/com/amazon/ion/benchmark/WriteRandomIonValues.java
Comment (linlin-s) [reply src/com/amazon/ion/benchmark/WriteRandomIonValues.java]: > You already have a `Random` instance, and should probably use that here instead of `Math.random()` (which creates its own `Random` instance). That should allow you to use `nextLong()` to avoid the cast.

To avoid this situation, the new updates in the following commit use `ThreadLocalRandom.current().nextLong(lowerBound, upperBound) ` to generate the random value which allows to set range.

File: src/com/amazon/ion/benchmark/WriteRandomIonValues.java
Comment (desaikd) [inline src/com/amazon/ion/benchmark/WriteRandomIonValues.java:396]:
```suggestion
     * Construct string which is conformed with the provided codepoint_length.
```

**Review by desaikd — APPROVED** [review]
Looks good! Some minor comments.

File: src/com/amazon/ion/benchmark/WriteRandomIonValues.java
Comment (linlin-s) [reply src/com/amazon/ion/benchmark/WriteRandomIonValues.java:363]: > 

Sounds good, I'll open an issue and add the link here.

**Pull request merged**
